notion-server/src/ethereum/mod.rs

```diff
@@ -1,25 +1,122 @@
+#![allow(unused)]
+
+use async_std::sync::{channel, Receiver, Sender};
+
+use snafu::{OptionExt, Snafu};
+
 use std::collections::HashMap;
 
+/// Shorthand for result types returned from the Simulation simulation.
+pub type Result<V, E = Error> = std::result::Result<V, E>;
+
+/// Errors arising from the simulation.
+#[derive(Debug, Snafu)]
+pub enum Error {
+    /// Operation was cancelled because the simulation is shutting down.
+    Terminated,
+}
+
+#[derive(Debug)]
+enum Operation {
+    CreateExecutionEnvironment(args::CreateExecutionEnvironment, Sender<EeIndex>),
+    CreateShardChain(args::CreateShardChain, Sender<u32>),
+    CreateShardBlock(args::CreateShardBlock, Sender<u32>),
+    GetShardBlock(args::GetShardBlock, Sender<ShardBlock>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Handle(Sender<Operation>);
+
+impl Handle {
+    pub async fn create_shard_chain(&self, arg: args::CreateShardChain) -> Result<u32> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::CreateShardChain(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn create_shard_block(&self, arg: args::CreateShardBlock) -> Result<u32> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::CreateShardBlock(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn create_execution_environment(
+        &self,
+        arg: args::CreateExecutionEnvironment,
+    ) -> Result<EeIndex> {
+        let (sender, receiver) = channel(1);
+
+        self.0
+            .send(Operation::CreateExecutionEnvironment(arg, sender))
+            .await;
+
+        receiver.recv().await.context(Terminated)
+    }
+
+    pub async fn shard_block(&self, arg: args::GetShardBlock) -> Result<ShardBlock> {
+        let (sender, receiver) = channel(1);
+
+        self.0.send(Operation::GetShardBlock(arg, sender)).await;
+
+        receiver.recv().await.context(Terminated)
+    }
+}
+
 #[derive(Debug)]
-pub struct EthereumSimulation {
+pub struct Simulation {
     beacon_chain: BeaconChain,
     shard_chains: Vec<ShardChain>,
+    receiver: Receiver<Operation>,
 }
 
-impl EthereumSimulation {
-    pub fn new() -> Self {
-        Self {
+impl Simulation {
+    pub fn new() -> (Self, Handle) {
+        let (sender, receiver) = channel(1);
+
+        let sim = Self {
             beacon_chain: BeaconChain::new(),
             shard_chains: Vec::new(),
+            receiver,
+        };
+
+        let handle = Handle(sender);
```

Comment:
Ahhh I see, this answers one of my questions up above, I missed this line earlier.  In this case, I'm curious still about the `recv` function on line 36 (and other similar lines in the other fns on Handler).  Can ask my follow-up questions offline, though.